File: benchmark/comparison/results/haiku_ts/10.ts
type Result<T> = { type: "Ok"; value: T } | { type: "Err"; message: string };

function validateUsername(username: string): Result<string> {
  if (username === "") {
    return { type: "Err", message: "username must not be empty" };
  }
  if (username.length > 20) {
    return { type: "Err", message: "username must be at most 20 characters" };
  }
  return { type: "Ok", value: username };
}

function validateEmail(email: string): Result<string> {
  if (!email.includes("@")) {
    return { type: "Err", message: "email must contain @" };
  }
  return { type: "Ok", value: email };
}

function validatePassword(password: string): Result<string> {
  if (password.length < 8) {
    return { type: "Err", message: "password must be at least 8 characters" };
  }
  return { type: "Ok", value: password };
}

function register(username: string, email: string, password: string): Result<string> {
  const usernameResult = validateUsername(username);
  if (usernameResult.type === "Err") {
    return usernameResult;
  }

  const emailResult = validateEmail(email);
  if (emailResult.type === "Err") {
    return emailResult;
  }

  const passwordResult = validatePassword(password);
  if (passwordResult.type === "Err") {
    return passwordResult;
  }

  return {
    type: "Ok",
    value: `${username} registered with ${email}`,
  };
}

function resultToString(result: Result<string>): string {
  if (result.type === "Ok") {
    return `Ok(${result.value})`;
  }
  return `Err(${result.message})`;
}

// Test cases
const testCases = [
  ["alice", "alice@example.com", "password123"],
  ["", "a@b", "pass1234"],
  ["alice", "noatsign", "pass1234"],
  ["alice", "a@b", "short"],
  ["aaaaabbbbbcccccddddde", "a@b", "pass1234"],
];

for (const [username, email, password] of testCases) {
  console.log(resultToString(register(username, email, password)));
}

File: benchmark/comparison/results/adv_py/10.py
def validate_username(s):
    if not s:
        return ("Err", "username must not be empty")
    if len(s) > 20:
        return ("Err", "username must be at most 20 characters")
    return ("Ok", s)

def validate_email(s):
    if "@" not in s:
        return ("Err", "email must contain @")
    return ("Ok", s)

def validate_password(s):
    if len(s) < 8:
        return ("Err", "password must be at least 8 characters")
    return ("Ok", s)

def register(u, e, p):
    r = validate_username(u)
    if r[0] == "Err":
        return f"Err({r[1]})"
    r = validate_email(e)
    if r[0] == "Err":
        return f"Err({r[1]})"
    r = validate_password(p)
    if r[0] == "Err":
        return f"Err({r[1]})"
    return f"Ok({u} registered with {e})"

tests = [
    ("alice", "alice@example.com", "password123"),
    ("", "a@b", "pass1234"),
    ("alice", "noatsign", "pass1234"),
    ("alice", "a@b", "short"),
    ("aaaaabbbbbcccccddddde", "a@b", "pass1234"),
]

for u, e, p in tests:
    print(register(u, e, p))
